### PR TITLE
Support GCP joining when `google` claim is not present

### LINKED ISSUE #28636: GCP Joining does not work with GKE Workload Identity
Expected behavior:

GCP joining should work with instances and SA from workload identity.

Current behavior:

GCP joining does not work with tokens generated from service accounts under GKE workload identity because they miss the `google` claim.

Teleport expects the JWT token to use the format 

```json
{
   "iss": "[TOKEN_ISSUER]",
   "iat": [ISSUED_TIME],
   "exp": [EXPIRED_TIME],
   "aud": "[AUDIENCE]",
   "sub": "[SUBJECT]",
   "azp": "[AUTHORIZED_PARTY]",
   "google": {
    "compute_engine": {
      "project_id": "[PROJECT_ID]",
      "project_number": [PROJECT_NUMBER],
      "zone": "[ZONE]",
      "instance_id": "[INSTANCE_ID]",
      "instance_name": "[INSTANCE_NAME]",
      "instance_creation_timestamp": [CREATION_TIMESTAMP],
      "instance_confidentiality": [INSTANCE_CONFIDENTIALITY],
      "license_id": [
        "[LICENSE_1]",
          ...
        "[LICENSE_N]"
      ]
    }
  }
}

``` 
To receive this token, Teleport does

```bash
curl -s "http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/token?audience=teleport.cluster.local&format=full&licenses=FALSE" --header "Metadata-Flavor: Google"
```

But for identities using the GKE workload, all tokens miss the "google" claim even when `format=full` is used, which results in pods being unable to join the Teleport cluster.

```json
 {
  "aud": "teleport.cluster.local",
  "azp": "115802695493397001654",
  "email": "[SA_NAME]@[PROJECT_ID].iam.gserviceaccount.com",
  "email_verified": true,
  "exp": 1688476552,
  "iat": 1688472952,
  "iss": "https://accounts.google.com",
  "sub": "115802695493397001654"
}

```

When the token presents an `email`, we should consider extracting the project id from it to match the Teleport Join token.

# Recreation steps

Follow https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity and 

```bash
curl -s "http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/token?audience=teleport.cluster.local&format=full&licenses=FALSE" --header "Metadata-Flavor: Google"
```

### PR DESCRIPTION
**Pull request by tigrato**
Teleport assumes that the `google` claim is present in the identity token that the Teleport service shares with Auth server. This is valid for VMs but it's not valid for GKE clusters using Workload identity and other GCP services. Teleport requests the identity token with `format=full` to receive this enhanced token.

Example of an identity token with a `google` claim:

```json
{
   "iss": "[TOKEN_ISSUER]",
   "iat": [ISSUED_TIME],
   "exp": [EXPIRED_TIME],
   "aud": "[AUDIENCE]",
   "sub": "[SUBJECT]",
   "azp": "[AUTHORIZED_PARTY]",
   "google": {
    "compute_engine": {
      "project_id": "[PROJECT_ID]",
      "project_number": [PROJECT_NUMBER],
      "zone": "[ZONE]",
      "instance_id": "[INSTANCE_ID]",
      "instance_name": "[INSTANCE_NAME]",
      "instance_creation_timestamp": [CREATION_TIMESTAMP],
      "instance_confidentiality": [INSTANCE_CONFIDENTIALITY],
    }
  }
}
```

The problem arises when one tries to use GCP joining for a Teleport service running on a GKE pod. When inside a pod with a binding between the Kubernetes Service account and the Google IAM Service Account, Google's metadata service token does not include the `google` claim. so it fails to join the cluster because of the unknown `project_id`.

```json
{
  "aud": "teleport.cluster.local",
  "azp": "115802695493397001654",
  "email": "<service_account_name>@<project_id>.iam.gserviceaccount.com",
  "email_verified": true,
  "exp": 1688476207,
  "iat": 1688472607,
  "iss": "https://accounts.google.com",
  "sub": "115802695493397001654"
}
```

To bypass this limitation, this PR extracts the `project_id` from the Google Service Account Email claim
`<service_account_name>@<project_id>.iam.gserviceaccount.com`. We use regex to extract the `project_id` and ensure the email follows the specified format above. Tests were introduced to validate the email.

Fixes #28636

### CODE REVIEW

File: lib/auth/join_gcp.go [hunk 1]
```diff
@@ -66,6 +66,9 @@ func checkGCPAllowRules(token *types.ProvisionTokenV2, claims *gcp.IDTokenClaims
 		if !slices.Contains(rule.ProjectIDs, compute.ProjectID) {
 			continue
 		}
+
+		// compute.Zone is not set if running in a GKE cluster under Workload Identity.
+		// In this case, `google` claim is not present and we can't check the zone.
 		if len(rule.Locations) > 0 && !slices.ContainsFunc(rule.Locations, func(location string) bool {
```
Comment (tigrato) [inline lib/auth/join_gcp.go:72]: What should we do here? 
`Locations` are never available without the `google` claim, so Pods won't be able to join if the location is specified in Teleport Token.

Should we skip the check if `compute.Zone==""`  or should we return an error saying: service account <ABC> didn't provide the location. Remove it from the token immediately

File: lib/gcp/token_validator.go [hunk 2]
```diff
@@ -92,5 +93,35 @@ func (id *IDTokenValidator) Validate(ctx context.Context, token string) (*IDToke
...
+	claims.Google.ComputeEngine.ProjectID = matches[1]
+
 	return &claims, nil
 }
+
+// gcpServiceAccountEmailRegex is the regex used to extract the project ID from
+// the email claim of a GCP service account.
+// See https://cloud.google.com/iam/docs/service-accounts#service_account_email_addresses
+// See https://cloud.google.com/iam/docs/service-account-types#user-managed
+var gcpServiceAccountEmailRegex = regexp.MustCompile(`(?s)^[^@]+@([^\.]+)\.iam\.gserviceaccount\.com$`)
```
Comment (strideynet) [inline lib/gcp/token_validator.go:127]: Can we name this to make it clear this is only for user managed service accounts? 

This won't match for the default compute service account which looks like `project-number-compute@developer.gserviceaccount.com`.

This also raises the question, should we support the use of the default compute service account ? It's definitely poor security practice for customers to be using this - but I wouldn't be surprised if someone tries it - perhaps we can log an error/warning that at least explains that we don't support the default compute service account ?

File: lib/auth/join_gcp.go (hunk 1 above)
Comment (strideynet) [reply lib/auth/join_gcp.go:72]: I'm leaning towards emitting a warning here and rejecting the join - it's easy enough for them to modify the token to add an allow rule that doesn't specify a zone and with rules around joining/authz, it seems better to lean towards being too conservative.

File: lib/gcp/token_validator.go [hunk 3]
```diff
@@ -92,5 +93,35 @@ func (id *IDTokenValidator) Validate(ctx context.Context, token string) (*IDToke
 	if err := idToken.Claims(&claims); err != nil {
 		return nil, trace.Wrap(err)
 	}
+
+	if claims.Google.ComputeEngine != (ComputeEngine{}) {
+		return &claims, nil
+	}
+
+	// If the token does not have a `google.compute_engine` claim, it is not a
```
Comment (strideynet) [inline lib/gcp/token_validator.go:101]: Nit: the first two sentences in this comment seem to explain the same thing twice.

File: lib/gcp/token_validator.go (hunk 2 above)
Comment (tigrato) [reply lib/gcp/token_validator.go]: I deliberately removed the support for the default compute service account.

Added an error rejecting the default service account usage

Comment (strideynet) [reply lib/gcp/token_validator.go]: Perfect - we want to guide people away from that anyway.

File: lib/gcp/token_validator.go [hunk 4]
```diff
@@ -120,8 +121,15 @@ func (id *IDTokenValidator) Validate(ctx context.Context, token string) (*IDToke
 	return &claims, nil
 }
 
-// gcpServiceAccountEmailRegex is the regex used to extract the project ID from
-// the email claim of a GCP service account.
-// See https://cloud.google.com/iam/docs/service-accounts#service_account_email_addresses
-// See https://cloud.google.com/iam/docs/service-account-types#user-managed
-var gcpServiceAccountEmailRegex = regexp.MustCompile(`(?s)^[^@]+@([^\.]+)\.iam\.gserviceaccount\.com$`)
+var (
+	// gcpManagedServiceAccountEmailRegex is the regex used to extract the project ID from
```
Comment (strideynet) [inline lib/gcp/token_validator.go:125]: Nit: I think without the word "user" this is potentially more confusing - these are SAs created and managed by users of GCP, not SAs managed by GCP (which is what the Default Compute SA is)

**Review by strideynet — APPROVED** [review]

File: lib/gcp/token_validator_test.go [hunk 5]
```diff
@@ -176,6 +175,119 @@ func TestIDTokenValidator_Validate(t *testing.T) {
...
+				Google: Google{
+					ComputeEngine: ComputeEngine{
+						ProjectID: "project-id",
+					},
+				},
+			},
+		},
+		{
+			name: "default service account: @developer.gserviceaccount.com domain",
+			assertError: func(tt require.TestingT, err error, i ...interface{}) {
```
Comment (jakule) [inline lib/gcp/token_validator_test.go:203]: nit:
```suggestion
			assertError: func(tt require.TestingT, err error, i ...any) {
```

File: lib/auth/join_gcp.go [hunk 6]
```diff
@@ -61,24 +65,36 @@ func (a *Server) checkGCPJoinRequest(ctx context.Context, req *types.RegisterUsi
...
 		}
 
 		// All provided rules met.
 		return nil
 	}
 
+	// If the location restriction is set and the "google.compute_engine.zone" claim is not present in the IDToken,
+	// return a more specific error message.
+	if unmatchedLocation && compute.Zone == "" {
+		return trace.CompareFailed("id token %q claim is empty and didn't match the %q. "+
```
Comment (jakule) [inline lib/auth/join_gcp.go:95]: Why here the format string is always used to insert the same value?

**Review by jakule — APPROVED** [review]

File: lib/auth/join_gcp.go (hunk 6 above)
Comment (tigrato) [reply lib/auth/join_gcp.go:95]: I used the format string to avoid having to escape `"`

Comment (atburke) [reply lib/auth/join_gcp.go:95]: You could also use a raw string (with `` ` ``) to do the same thing.

File: lib/gcp/token_validator.go (hunk 2 above)
Comment (atburke) [reply lib/gcp/token_validator.go]: @tigrato If we're removing support for the default service account, we'll need to update the docs to reflect that.

**Review by atburke — APPROVED** [review]

File: lib/gcp/token_validator.go (hunk 2 above)
Comment (tigrato) [reply lib/gcp/token_validator.go]: @atburke I only removed the support if it's not running in a VM  but I can document it 

@strideynet should we also consider removing support from VM instances? I am afraid that someone else already uses it

**Comment by multani** [comment]
Another possibility would be to relax the mandatory "project ID" in the token definition: if the project can be extracted from the service account email address, having both of them configured in the token definition doesn't really improve the security, and having only the service account email would be the same, no?

In which case:

* If a project ID is configured in the token: try to match it against a claim and fail if not available
* if a service account is configured, either a default one or a user one, match it against a claim and fail otherwise
* if both are configured, match against all the available claim and fail if no match

So no need to extract anything, the default service accounts can still be used and it's up to the Teleport operator to configure the token definition so that it can match the claims provided by the ID token from GCP.

I may have misunderstood the token options for GCP in Teleport and how they all work together (I couldn't find a good documentation about it, except form the example in https://goteleport.com/docs/agents/join-services-to-your-cluster/gcp/#step-13-create-the-gcp-joining-token)

**Comment by tigrato** [comment]
@multani 

The idea behind enforcing the `project_id` in the Teleport token is to ensure that customers restrict access to their own projects.  Google uses the same JWT signing keys for `accounts.google.com` which creates the possibility of having random Google accounts be able to join your cluster if the `project_id` wasn't mandatory.  We can relax and allow at least one `service_account` or `project_id` to be present but it's clear to enforce the `project_id` and derive the value from the SA email when the `google` claim isn't present.

The field `service_accounts` aims to add an additional restriction layer on top of `project_id`. The idea is to be able to differentiate the roles that each service account can assume - i.e. having a service account that can join as a node while others can only join as Kube.


```yaml
kind: token
spec:
  # use the minimal set of roles required (e.g. Node, Proxy, App, Kube, DB, WindowsDesktop)
  roles: [Node]

  # set the join method allowed for this token
  join_method: gcp

  gcp:
    allow:
      # The GCP project ID(s) that VMs can join from.
      - project_ids: ["example-project-id"]
        # (Optional) The email addresses of service accounts that VMs can join
        # with.
        service_accounts: ["vm-sa@<project-id>.iam..."]

```
```yaml
kind: token
  roles: [Kube]
  # set the join method allowed for this token
  join_method: gcp
  gcp:
    allow:
      # The GCP project ID(s) that VMs can join from.
      - project_ids: ["example-project-id"]
        # (Optional) The email addresses of service accounts that VMs can join
        # with.
        service_accounts: ["kube-sa@<project-id>.iam..."]

```

> the default service accounts can still be used and it's up to the Teleport operator to configure the token definition so that it can match the claims provided by the ID token from GCP.

I deliberately excluded the default service account for security reasons. There is no other reason than that. It is possible to extract the `project_id` from the default account `<project_id>@developer.gserviceaccount.com` but we must encourage better security practices.


Thank you for your suggestion

**Pull request merged**